service/history/historyEngine.go

```diff
@@ -158,12 +166,63 @@ func (e *historyEngineImpl) StartWorkflowExecution(startRequest *h.StartWorkflow
 		return nil, &workflow.BadRequestError{Message: "Missing or invalid TaskStartToCloseTimeoutSeconds."}
 	}
 
-	// We generate a new workflow execution run_id on each StartWorkflowExecution call.  This generated run_id is
-	// returned back to the caller as the response to StartWorkflowExecution.
-	runID := uuid.New()
-	workflowExecution := workflow.WorkflowExecution{
-		WorkflowId: common.StringPtr(executionID),
-		RunId:      common.StringPtr(runID),
+	execution := workflow.WorkflowExecution{
+		WorkflowId: common.StringPtr(*request.WorkflowId),
+	}
+	isBrandNewWorkflow := true
+	// without setting the run ID so we can get the current workflow. if any
+	context, release, err := e.historyCache.getOrCreateWorkflowExecution(domainID, execution)
+	execution.RunId = common.StringPtr(uuid.New())
+
+	errFn := func(errMsg string, ms *mutableStateBuilder) error {
+		msg := fmt.Sprintf(errMsg, ms.executionInfo.WorkflowID, ms.executionInfo.RunID)
+		return &workflow.WorkflowExecutionAlreadyStartedError{
+			Message:        common.StringPtr(msg),
+			StartRequestId: common.StringPtr(fmt.Sprintf("%v", ms.executionInfo.CreateRequestID)),
+			RunId:          common.StringPtr(fmt.Sprintf("%v", ms.executionInfo.RunID)),
+		}
+	}
+
+	if err != nil {
+		if _, ok := err.(*workflow.EntityNotExistsError); !ok {
+			return nil, err
+		}
+		// here we know that there is no existing run of this workflow ID, i.e. not running, no history.
+	} else {
+		isBrandNewWorkflow = false
+		defer release()
+
+		msBuilder, err := context.loadWorkflowExecution()
+		if err != nil {
+			return nil, err
+		}
+		// here we know there is some information about the workflow, i.e. either running right now or has history
+		// check if the this workflow is finished
+		if msBuilder.isWorkflowExecutionRunning() {
+			// if client issue a duplicate request
+			if msBuilder.executionInfo.CreateRequestID == common.StringDefault(request.RequestId) {
+				return &workflow.StartWorkflowExecutionResponse{RunId: common.StringPtr(msBuilder.executionInfo.RunID)}, nil
+			}
+			msg := "Workflow execution already running. WorkflowId: %v, RunId: %v."
+			return nil, errFn(msg, msBuilder)
+		}
+
+	StartType:
+		switch startRequest.StartRequest.GetWorkflowIdReusePolicy() {
+		case workflow.WorkflowIdReusePolicyAllowDuplicateFailedOnly:
+			if _, ok := FailedWorkflowCloseState[msBuilder.executionInfo.CloseStatus]; !ok {
+				msg := "Workflow execution already finished successfully. WorkflowId: %v, RunId: %v."
+				return nil, errFn(msg, msBuilder)
+			}
+			break StartType
+		case workflow.WorkflowIdReusePolicyAllowDuplicate:
+			break StartType
+		case workflow.WorkflowIdReusePolicyRejectDuplicate:
+			msg := "Workflow execution already finished. WorkflowId: %v, RunId: %v."
+			return nil, errFn(msg, msBuilder)
+		default:
+			return nil, &workflow.InternalServiceError{Message: "Internal Server Error"}
```

Comment:
need better error message.

Reply:
ok

Reply:
as long as the logging framework correctly log this error, then we have enough information to debug

Reply:
checked history service, do not see any err logging done, so updated the error message